### ISSUE TITLE
Planned feature: instantaneous bottlenecks with non-binary trees

### ISSUE THREAD

**Issue by jeromekelleher** [tskit-dev/msprime#9]
Implement a new DemographicEvent class that allows us to model instantaneous bottlenecks as described in [Galtier et al. 2010](http://www.genetics.org/content/155/2/981.short) and [Bunnefeld et al 2015](http://www.genetics.org/content/201/3/1157.short).

This will also require us to model non-binary trees, which will need require some low-level replumbing, as well as a change to the HDF5 file format.

**Comment by KLohse** [comment]
Yes, this would be a useful feature. Assuming an instantaneous event is the simplest model of a bottleneck. Allowing for non-binary trees may also make it possible to use msprime to simulate more general multiple merger coalescents in the future.

**Comment by jeromekelleher** [comment]
The main code for this has just been merged into master. There is an undocumented demographic event ``msprime.Bottleneck(time, population_id, proportion)`` which merges a fraction p of the lineages in a population at a given time. This doesn't follow any established population genetic models, but is useful for testing and development. Hopefully we can follow this up with some demographic events that implement the Galtier et al model, and also add in recurrent k-merger events for Lambda coalescent support (@shajoezhu is looking into this).

All of the TreeSequence and API natively supports non binary trees now.

**Comment by jeromekelleher** [comment]
I've finally gotten around to implementing this @KLohse. I have an initial pass at the implementation in the  ``instantaneous_bottlenecks`` branch, which you can check out and try if you want to take it for a spin. Here is an example of using the code:

```python
    demographic_events = [
        msprime.InstantaneousBottleneck(time=1000, strength=1e4)]
    ts = msprime.simulate(
            sample_size=10, Ne=1e4, recombination_rate=1e-5,
            length=10, demographic_events=demographic_events)
    for record in ts.records():
        print(
            "{:.2f}-{:.2f}\t{:.1f}".format(
                record.left, record.right, record.time),
            record.node, record.children, sep="\t")
```
The ``strength`` parameter here is equivalent to T2 in the usual terminology and is measured in *generations*. In this case, we create a bottleneck equivalent to 10K generations at generation 1000. The output we get is

```
5.74-10.00      528.5   10      (5, 6)
0.00-5.74       1000.0  11      (6, 8)
0.00-10.00      1000.0  12      (1, 2, 3, 4, 7, 9)
3.02-3.68       8256.1  13      (0, 12)
3.68-5.75       9603.1  14      (0, 12)
9.68-10.00      9797.0  15      (0, 12)
0.00-0.90       10362.9 16      (0, 12)
0.00-5.74       14981.3 17      (5, 11)
5.74-6.54       14981.3 17      (8, 10)
0.90-3.02       19780.1 18      (0, 12)
6.11-6.54       22698.5 19      (0, 17)
6.54-9.68       22698.5 19      (0, 10)
6.54-9.68       28214.7 20      (8, 12)
9.68-10.00      28214.7 20      (8, 15)
0.00-0.90       38687.4 21      (16, 17)
0.90-3.02       38687.4 21      (17, 18)
3.02-3.68       38687.4 21      (13, 17)
3.68-5.75       38687.4 21      (14, 17)
5.75-6.11       38687.4 21      (0, 17)
5.75-6.11       44471.8 22      (12, 21)
6.11-6.54       44471.8 22      (12, 19)
6.54-9.68       44471.8 22      (19, 20)
9.68-10.00      44471.8 22      (10, 20)
```

Note that records 2 and 3 are for different nodes, occur at the same time and can have lots of children. This is the signature of multiple simultaneous coalescences.

There are a few remaining issues:

1. This is nowhere near ready for release and is not tested properly. It may blow up.
2. I've limited the implementation to only work on a single population right now. What should happen when we have multiple populations? Is everything dumped into the same pot, or does each population have it's own independent bottleneck process?
3. In a similar vein, how do we deal with population growth and so on? At the moment I think we're assuming that we get N generations of the coalescent *under the current paremeters*. I.e., if the growth rate has been set to non-zero, then this will affect the bottleneck also.
4. Is ``InstantaneousBottleneck`` a good name? I'm using ``SimpleBottleneck`` for the case where we just merge a fraction ``p`` of the ancestors. Also, is ``strength`` the right name for the parameter?

If you want to try the code out, use the following:
```
$ virtualenv msprime-bottlenecks
$ source msprime-bottlenecks/bin/activate
$ git clone https://github.com/jeromekelleher/msprime.git
$ cd msprime
$ git checkout instantaneous-bottleneck
$ python setup.py install
```

This will avoid clobbering a working msprime install.

**Comment by jeromekelleher** [comment]
I've beefed up the tests a bit on this and merged it into master @KLohse. You should be able to use it now by skipping the ``git checkout instantaneous-bottleneck`` step above.

We'll need to do some more testing to make sure it meets the analytical expectations and document it also before closing this issue.

**Comment by KLohse** [comment]
Hi Jerome,

I am testing the bottleneck implementation right now. Will check against
analytic results for the SFS and blockwise SFS and let you know ASAP.

- I think "instantaneous bottleneck" is a good name, that's what Nick
called it in the Galtier et al. paper.

- It may be sensible to have the bottleneck affect all pops by default, but
I think being able to restrict it to a specific pop (or set of pops) is
important. For example one, might be interested in an instantaneous
bottleneck immediately following a split between populations.

-  Making the bottleneck relative to the current N_e as you say makes a lot
of sense to me. That way it wouldn't affect other demographic processes,
like continuous growth.

- Yup, I think strength is the right name for T2.

More soon,

Konrad

There are a few remaining issues:

   1. This is nowhere near ready for release and is not tested properly. It
   may blow up.
   2. I've limited the implementation to only work on a single population
   right now. What should happen when we have multiple populations? Is
   everything dumped into the same pot, or does each population have it's own
   independent bottleneck process?
   3. In a similar vein, how do we deal with population growth and so on?
   At the moment I think we're assuming that we get N generations of the
   coalescent *under the current paremeters*. I.e., if the growth rate has
   been set to non-zero, then this will affect the bottleneck also.
   4. Is InstantaneousBottleneck a good name? I'm using SimpleBottleneck
   for the case where we just merge a fraction p of the ancestors. Also, is
   strength the right name for the parameter?

**Comment by jeromekelleher** [comment]
Excellent, thanks @KLohse! I've updated the implementation to allow more than one population. All bottlenecks happen in exactly one population: if you want to mix the populations together before the bottleneck, you can use a MassMigration event just before it. This simplifies the semantics of the event quite a bit and makes it as similar as possible to the other types of events.

Please send me on the code for the analytical checks and I'll integrate them into the statistical test suite. After that, we just have to document the events and it's ready to release I think (assuming the statistical tests pass).

**Comment by KLohse** [comment]
Hi Jerome,

All seems fine. Comparing the analytic prediction for the probability
blockwise configurations defined via the SFS (counts of singleton,
doubletons, etc in blocks of a fixed length) to msprime (1mio sims), I get
a very good fit (see attached). I've only looked at one arbitrary point in
parameter space: theta 1.6 T1= 0.25 and T2 (strength)=0.5, but it would be
a huge coincidence I think if it matches

The code is in a little script, the actual comparison is done in
Mathematica. If it's easier to do the plotting in python I can give you the
analytic results as a table.

Cheers,
Konrad

**Comment by jeromekelleher** [comment]
GitHub doesn't send attachments @KLohse --- I guess you're using email interface? Just email the stuff directly to me. Thanks!

**Comment by stsmall** [comment]
Hi Jerome and Konrad,
Is it currently advisable to use this instantaneous bottleneck function in msprime? Also, are there functions available in msprime to calculate the bSFS and various higher-level SFS for comparison to observed data?
thank you,
scott

**Comment by jeromekelleher** [comment]
Apologies for the slow response Scott, I've been on vacation. I think Konrad is also, but I'm not sure.

I think we're pretty happy that this code is correct @stsmall, we're basically just waiting on @KLohse to produce some documentation and examples (hint, hint Konrad!). If the examples above seem clear enough to you, please go ahead and use it. If you have any feedback at all, please let us know.

There aren't any high-level functions for computing SFS related objects in msprime currently, but this is something I'm sure would be generally useful. It's not on my radar at the moment, but if you would like to help and chip in some code I'd be happy to point in the right direction.

**Comment by stsmall** [comment]
Thanks! Maybe Konrad has something? I can try to script something, probably try to borrow some methods from the dadi program (Gutenkunst) which uses numpy arrays to store high dimensional AFS. I am writing a script for dadi to use msprime, so I will try to also do the reverse.
I contacted Lynsey Bunnefeld about bSFS python code, I will ask her permission to adopt these into msprime ... since Konrad is her collaborator maybe he has that all worked out anyways, ?

**Comment by jeromekelleher** [comment]
Konrad definitely has some code for computing bSFS stuff, which I guess we should try and bring into msprime. I think he's still on vacation though, so is probably out of contact. I dug around in my emails and found this code that Konrad and I worked out a while back:
```python
import msprime
import numpy as np

"""
Exploratory code to compute the blockwise SFS.
"""

def sfs_branch_weights(tree):
    x = np.zeros(tree.get_sample_size())
    for u in tree.nodes():
        if tree.get_parent(u) != msprime.NULL_NODE:
            x[tree.get_num_tracked_leaves(u)] += tree.get_branch_length(u)
    return x

def block_sfs(ts, num_populations):
    x = np.zeros((num_populations, ts.get_sample_size()))
    for j in range(num_populations):
        for t in ts.trees(tracked_leaves=ts.get_samples(j)):
            x[j] += t.get_length() * sfs_branch_weights(t)
    return x

def main():
    population_configurations = [
        msprime.PopulationConfiguration(sample_size=2),
        msprime.PopulationConfiguration(sample_size=4),
    ]
    ts = msprime.simulate(
            recombination_rate=1,
            population_configurations=population_configurations,
            migration_matrix=[[0, 1], [1, 0]])
    print(block_sfs(ts, 2))

if __name__ == "__main__":
    main()
```
This might be a helpful starting point.

It's great to hear you're updating dadi to use msprime; let me know how it goes!

**Comment by KLohse** [comment]
Hi Scott, I just got back from travels, so apologies for my slow reply (and thanks Jerome for being quicker). 

I am afraid that I don't have much more in terms of SFS and bSFS code than the snipped Jerome posted. I've mainly used msprime to check analytic predictions and to set up parametric bootstrap reps. For the latter I have written some python scripts that chop up contiguous sequences (e.g. msprime simulations with mutations) into blocks of a fixed length and compute the bSFS. These should analogous to the bSFS python scripts you have from Lynsey but based on msprime rather than ms and I am happy to send them along if they would be of any help. In general, I completely agree with Jerome: some general code for computing bSFS and SFS related things from msprime simulations would be very useful indeed!

**Comment by champost** [comment]
Hi @stsmall @jeromekelleher @KLohse Just my two cents here on `ms` styled simulation output from `msprime` or other...

Once data is already in `DaDi`, the SFS can be manipulated easily in various ways. As to importing simulations in classic (binary) `ms` format into `DaDi`, you can use the function `dadi.Spectrum.from_ms_file()`.

As for the bSFS or joint-bSFS (jbSFS), you can use the `convert_data_to_bSFS` keyword in [ABLE](https://github.com/champost/ABLE) for converting any `ms` styled data/simulations in genotype/binary format respectively (for more see [Pg.6 of helpABLE](https://github.com/champost/ABLE/blob/master/doc/helpABLE.pdf)).

**Comment by stsmall** [comment]
Hi @champost, great suggestions! My motivation behind adding msprime to dadi was to keep everything in python rather than calling ms with popen. It is quite easy to convert msprime to SFS objects used by dadi, especially mimicking the functions in dadi.Spectrum.from_ms_file() as you pointed out. As for the bSFS, ABLE seems like a great program but do I then need a config file for each ms-type binary I want to transform into bSFS?

**Comment by stsmall** [comment]
Hi @KLohse,
I lost contact with Lynsey, does the offer still stand to provide your python scripts for use with msprime as mentioned above? you can email me off list at my gmail, stsmall.
thank you!

**Comment by jeromekelleher** [comment]
@KLohse --- going to nag you again here to get this done. We just want a statistical test to make sure it's correct and a little documentation...
